### PR TITLE
#1376 Register and cancel event sending fix

### LINKED ISSUE #1376: EurekaInstanceRegisteredEvent not dispatching as intended
I went through the source code and figured out that eureka dispatches the following 3 events from springs `ApplicationContext`

1. `EurekaInstanceRenewedEvent`
1. `EurekaInstanceRegisteredEvent`
1. `EurekaInstanceCancelledEvent`  
  

Of these only the EurekaInstanceRenewedEvent works and the listener is triggered when the server receives a heartbeat. Other events are not dispatched when a new instance is registered or canceled from eureka.

I think `InstanceRegistry` is overriding the wrong `register` method.  

The `ApplicationResource` controller calls the `register` method in the  `InstanceRegistry` when a new instance is registered. But the call goes to the super class of `InstanceRegistry`, bypassing the event dispatch. Not sure if this is intended and this is how it is supposed to work. If it is then what is the use of the event and how can it be used?

### PR DESCRIPTION
**Pull request by bslota**
#1376 This commit fixes application event publishing when an instance is being registered/canceled.

### CODE REVIEW

File: spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java [hunk 1]
```diff
@@ -126,4 +123,41 @@ public boolean renew(final String appName, final String serverId,
...
+	@Override
+	protected boolean internalCancel(String appName, String id, boolean isReplication) {
+		logCancelation(appName, id, isReplication);
+		publishEurekaInstanceCanceledEvent(appName, id, isReplication);
+		return super.internalCancel(appName, id, isReplication);
+	}
+
+	private void logRegistration(InstanceInfo info, boolean isReplication, int instanceLeaseDuration) {
+		if (log.isDebugEnabled()) {
+      log.debug("register " + info.getAppName() + ", vip " + info.getVIPAddress()
```
Comment (ryanjbaxter) [inline spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java]: spacing looks a little off here

Comment (bslota) [inline spring-cloud-netflix-eureka-server/src/main/java/org/springframework/cloud/netflix/eureka/server/InstanceRegistry.java]: Thanks for your feedback, I will fix formatting, and add tests

**Comment by bslota** [comment]
Tests of overriden methods end with the calls of netflix's superclass methods (`internalCancel `and `register`). In order to mock them, I wrapped them in `super*` methods. In tests, I'm checking if event is properly produced, and if mentioned wrappers are being invoked.

**Comment by ryanjbaxter** [comment]
@bslota Instead of creating methods you mock, have you considered using mockito and mocking the listener so you can verify which methods are called?  I am not really fond of adding code to a class for testing purposes.

**Comment by bslota** [comment]
@ryanjbaxter Are you referring to the test class? Or are you talking about `superRegister` and `superInternalCancel` methods? I must say I don't like the idea of wrapping super class methods at all - if you have any idea of how to mock superclass methods, I will appreciate any advice..

**Comment by ryanjbaxter** [comment]
I dont think u need to worry about what happens in the super calls from a unit test perspective.  You can assume that the super class is being tested properly.  You just need to worry about testing the code you added, which is essentially making sure the correct events get fired.

**Comment by bslota** [comment]
Have no idea why spring-cloud-netflix-core module test failed...

**Comment by bslota** [comment]
I tried to remove as many duplications as possible. I also used used created private methods in `renew`  method which initially wasn't changed, but I wanted the class to be consistent. Of course I added a test to check if `renew` also creates a proper event.

**Pull request merged**